java/src/main/java/linalg/algorithms/Householder.java

```diff
@@ -0,0 +1,122 @@
+package linalg.algorithms;
+
+import linalg.matrix.DenseMatrix;
+import linalg.matrix.DenseMatrixBuilder;
+import linalg.matrix.Matrix;
+import linalg.matrix.MatrixFactory;
+import linalg.vector.DenseVectorBuilder;
+
+import static linalg.matrix.MatrixFactory.Add;
+import static linalg.matrix.MatrixFactory.DenseMultiply;
+import static linalg.vector.VectorUtils.VectorLength;
+
+public class Householder implements QRDecomposition {
+
+  private Matrix Q;
+  private Matrix R;
+
+  @Override
+  public Matrix GetQ() {
+    return Q;
+  }
+
+  @Override
+  public Matrix GetR() {
+    return R;
+  }
+
+  @Override
+  public void Decompose(Matrix matrix) {
+    HouseholderAlgorithm(matrix);
+    R = MatrixFactory.DenseMultiply(Q.Transpose(), matrix);
+  }
+
+  private void HouseholderAlgorithm(Matrix A) {
+    if (A.Cols() == 0)
+      return;
+
+    DenseVectorBuilder a_vec_builder = FirstColumn(A);
+
+    a_vec_builder.SetValue(0, a_vec_builder.GetValue(0) +
+      VectorLength(a_vec_builder.BuildVector()) * Math.signum(a_vec_builder.GetValue(0)));
+
+    Matrix identity = IdentityMatrix(a_vec_builder.VectorSize());
+    Matrix fraction = FractionOfVectors(a_vec_builder);
+
+    Matrix H = Add(identity, fraction);
+    Q = (Q == null) ? H : DenseMultiply(Q, UnitColRows(H));
+
+    if (A.Cols() > 1 && A.Rows() > 1)
+      HouseholderAlgorithm(CutFirstColRow(DenseMultiply(H, A)));
+  }
+
+  private DenseVectorBuilder FirstColumn(Matrix matrix) {
+    if (matrix == null) {
+      throw new IllegalArgumentException("Matrix cannot be null");
+    }
+
+    DenseVectorBuilder builder = new DenseVectorBuilder(matrix.Rows());
+
+    for (int i = 0; i < builder.VectorSize(); i++) {
+      builder.SetValue(i, matrix.ValueAt(i, 0));
+    }
+
+    return builder;
+  }
+
+  private Matrix IdentityMatrix(int size) {
```

Comment:
THis kind of method shouldn't be inside HouseHolder. It's not HouseHolder specific. Please move it to MatrixFactory, like public static Matrix Identity(int size)